lib/ci-helper.ts

```diff
@@ -763,6 +772,211 @@ export class CIHelper {
+                                    name
+                                    user {
+                                        login
+                                    }
+                                }
+                            }
+                        }
+                    },
+                    url
+                    files(last:20) {
+                        totalCount,
+                        nodes {
+                            additions,
+                            deletions,
+                            path
+                        }
+                    }
+                }
+            }
+        }`;
+
+        type queryResponse = {
+            repository: Repository;
+        }
+
+        const repository =
+            await this.github.graphql<queryResponse>(
+            pr.baseOwner,
+            query
+        );
+
+        if (!repository || !repository.repository.pullRequest) {
+            throw new Error(`Pull request not found.`);
+        }
+        const prInfo = repository.repository.pullRequest;
+
+        if (prInfo.closed) {
+            return false;           // nothing to see here
+        }
+
+        const genLabel = (label: string, message: string): labelInfo => {
+            return {label, message};
+        };
+
+        try {
+            const commits = prInfo.commits.nodes;
+
+            if (commits) {
+                commits.map(node => {
+                    const login = node?.commit.committer?.user?.login ??
+                        node?.commit.author?.user?.login;
+
+                    // login may not be present if email unknown to github
+                    if (login && pr.author !== login) {
+                        throw new Error("Request author is not commit author.");
+                    }
+                });
+            }
+
+            const files = prInfo.files;
+            if (files && files.totalCount && files.nodes) {
+                files.nodes.map(node => {
+                    const changedFile = node as PullRequestChangedFile;
+                    this.checkNewUserFile(changedFile, files.totalCount,
+                        genLabel);
+                });
+            } else {
+                throw new Error("No files updated.");
+            }
+
+            const queryRef = `query {
+                repository(name: "${pr.baseRepo}", owner: "${pr.author}") {
+                    ref(qualifiedName: "${prInfo.baseRefName}") {
+                        name
+                    }
+                }
+            }`;
+
+            const refInfo = await this.github.graphql<queryResponse>(
+                pr.baseOwner,
+                queryRef
+            );
+
+            if (!refInfo || !refInfo.repository.ref) {
+                throw new Error(`Base branch ${prInfo.baseRefName
+                    } not found in ${pr.author}/${pr.baseRepo}.`);
+            }
+        } catch (e) {
+            if ((e as labelInfo).label === undefined ) {
```

Comment:
Does that mean that the absence of a label should tell us that the PR should be closed? Maybe an explicit `.closePR: boolean` would be better?

Reply:
The code is throwing an `Error` for PRs to be closed and a `labelInfo` for PRs to be flagged.  The check for `label` is really a type check (is there another way to do it?).  If you want, a single type can be used that has `label?`, `closePR` and `message`.

Reply:
I think that would be better. The current version looks a little bit as if it tries too hard to imitate Java's `instanceof` operator (or C++'s RTTI), which Javascript does not have.